[stdlib] Significant UTF8/16 decode speed-ups for iterator nil-guarantee

PatrickPijnappel opened the pull request:
<!-- Please complete this template before creating the pull request. -->
#### What's in this pull request?
<!-- Description about pull request. -->
Refactors `UTF8.decode(_:)` and `UTF16.decode(_:)` mainly for speed-ups based on the iterator-nil guarantee introduced in [SE-0052](https://github.com/apple/swift-evolution/blob/master/proposals/0052-iterator-post-nil-guarantee.md). `UTF32.decode(_:)` was slightly refactored for consistency.
- [Private benchmarks](https://gist.github.com/PatrickPijnappel/7cd42822330a1191bb57adff24c3085c) show speedups of ~20% for UTF-8 ASCII, ~10% for UTF-8 mixed Unicode, and ~40% for UTF-16, both for ASCII & mixed. Compiled using the 06/06 (a) snapshot with `-O -whole-module-optimization`. 
- I haven't been able to add a benchmark for UTF-16 as there seems to be a build failure after regenerating the harness files on master ([see here](http://article.gmane.org/gmane.comp.lang.swift.devel/2134/match=harness)).
- Verified against original implementation for all 32-bit sequences.

* * * *

<!-- This selection should only be completed by Swift admin -->
Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please smoke test
        All supported platforms     | @swift-ci Please smoke test and merge
        OS X platform               | @swift-ci Please smoke test OS X platform
        Linux platform              | @swift-ci Please smoke test Linux platform

**Validation Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please test
        All supported platforms     | @swift-ci Please test and merge
        OS X platform               | @swift-ci Please test OS X platform
        OS X platform               | @swift-ci Please benchmark
        Linux platform              | @swift-ci Please test Linux platform


**Lint Testing**

        Language     | Comment
        ------------ | -------------
        Python       | @swift-ci Please Python lint

Note: Only members of the Apple organization can trigger swift-ci.
</details>
<!-- Thank you for your contribution to Swift! -->

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@swift-ci Please benchmark

gribozavr commented on stdlib/public/core/Unicode.swift:
```diff
@@ -451,14 +440,7 @@ public struct UTF16 : UnicodeCodec {
-  ///
-  /// `x` is set when `_decodeLookahead` contains a code unit.
-  internal var _lookaheadFlags: UInt8 = 0
+  internal var _decodeLookahead: UInt32?
```
Have you thought about using something like `UInt32.max` for the `nil` value?

PatrickPijnappel replied:
`Uint32.max` is a valid value here.

gribozavr replied:
But isn't this UTF-16 decoding?  I think this buffer only stores UTF-16 code units, and nothing greater than UInt16.max will be stored in it.

PatrickPijnappel replied:
@gribozavr My bad, you're right. Would there be a difference though on 64-bit systems? 

Perhaps it's more consistent to make it a `UInt16?` or `CodePoint?`.

gribozavr replied:
`UInt32` seems to be convenient for the code below, using `UInt16` would require a truncation in one place.  This was just a thought.  Not necessarily actionable.

PatrickPijnappel replied:
It doesn't require any truncation, both version have the same number of (widening only) casts. I slightly prefer the UInt16 version, updated.

gribozavr commented:
Thank you!  PR LGTM, let's see what CI says.

gribozavr commented:
@PatrickPijnappel Could you take a look at CI failures?

PatrickPijnappel commented:
@gribozavr I'll take a look at the tests & regressing benchmarks.

jckarter commented:
The benchmark bot is a bit noisy. It's probably worth doing a few runs before jumping to conclusions.

gribozavr commented:
@swift-ci Please benchmark

PatrickPijnappel commented:
@gribozavr There were some validation tests that were explicitly checking for not calling `.next()` after it had returned `nil`, I removed those – can you rerun the tests?

I'll look into the benchmarks next.

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@swift-ci Please benchmark

gribozavr commented:
@swift-ci Please benchmark

gribozavr commented:
@swift-ci Please smoke test Linux platform

gribozavr commented:
@swift-ci Please benchmark

aschwaighofer commented:
I took a look at StringHasPrefix -- it is just very noisy and I think we can ignore it (it tests the ascii string fast path of String.hasPrefix and the change should not really affect it) and I got similar results before/after the change.
I also looked at OpenClose and the generated code for the hot function is the same yet there is a reproducible difference - we can ignore this. 
I did not look at Walsh but it should not be affected by the change other than noise.

I think we can merge this from a performance perspective. I have not looked at the linux platform regression. Dmitri should give the final okay or perform the merge.

gribozavr commented:
@aschwaighofer Thank you very much for the review!  This change LGTM.  The CI on master is red currently, I will merge when it improves.

PatrickPijnappel commented:
I added a [gist link](https://gist.github.com/PatrickPijnappel/7cd42822330a1191bb57adff24c3085c) above for the private benchmark I used. I refactored it to include the `UnicodeScalar(_unchecked:)` gains for UTF-16, which now in total appears to be a ~40% speed-up on both ASCII & mixed Unicode.

I haven't been able to identify why the UTF-8 gains aren't showing up in the formal UTF-8 benchmark though.

gribozavr commented:
@swift-ci Please test and merge

gribozavr commented:
@PatrickPijnappel Could you contribute your benchmark to the benchmark suite?

gribozavr commented:
@swift-ci Please test and merge

PatrickPijnappel commented:
@gribozavr Linux benchmark was still failing...

gribozavr commented:
@swift-ci Please benchmark

gribozavr commented:
@swift-ci Please benchmark

gribozavr commented:
@swift-ci Please test

gribozavr commented:
The benchmark regressions look like noise to me.
